cpp/locality/NeighborComputeFunctional.h

```diff
@@ -0,0 +1,136 @@
+#ifndef NEIGHBOR_COMPUTE_FUNCTIONAL_H
+#define NEIGHBOR_COMPUTE_FUNCTIONAL_H
+
+#include <tbb/tbb.h>
+
+#include "NeighborQuery.h"
+#include "AABBQuery.h"
+#include "NeighborList.h"
+#include "Index1D.h"
+
+namespace freud { namespace locality {
+
+class RawPoints : public NeighborQuery
+    {
+    public:
+        RawPoints();
+
+        RawPoints(const box::Box &box, const vec3<float> *ref_points, unsigned int Nref) :
+            NeighborQuery(box, ref_points, Nref) {}
+
+        ~RawPoints() {}
+
+        // dummy implementation for pure virtual function in the parent class
+        virtual std::shared_ptr<NeighborQueryIterator> query(const vec3<float> *points, unsigned int N, unsigned int k, bool exclude_ii=false) const
+        {
+            return nullptr;
+        }
+
+        // dummy implementation for pure virtual function in the parent class
+        virtual std::shared_ptr<NeighborQueryIterator> queryBall(const vec3<float> *points, unsigned int N, float r, bool exclude_ii=false) const
+        {
+            return nullptr;
+        }
+
+    };
+
+// ComputePairType should be a void function that takes (ref_point, point) as input.
+template<typename ComputePairType>
+void loop_over_NeighborList(const NeighborQuery *ref_points, const vec3<float> *points, unsigned int Np,
+                                  QueryArgs qargs, const NeighborList* nlist, const ComputePairType& cf)
+    {
+    // check if nlist exists
+    if(nlist != NULL)
+        {
+
+        // if nlist exists, loop over it parallely
+        const size_t *neighbor_list(nlist->getNeighbors());
+        size_t n_bonds = nlist->getNumBonds();
+        parallel_for(tbb::blocked_range<size_t>(0, n_bonds),
+            [=] (const tbb::blocked_range<size_t>& r)
+            {
+                for(size_t bond = r.begin(); bond !=r.end(); ++bond)
+                    {
+                    size_t i(neighbor_list[2*bond]);
+                    size_t j(neighbor_list[2*bond + 1]);
+                    cf(i, j);
```

Comment:
Is it possible to include neighbor distances and weights here? `cf(i, j, dist, weight);`?

Something like `auto dist = nq->getBox().wrap(ref_points->getRefPoints()[i] - points[j]);`
(make sure I put `i` and `j` in the right order, etc.)

Adding `distance`s will make it much easier to write the actual compute functions.

We don't currently use `weight`s to my knowledge but I want to run some analyses that incorporate weights so this is a good time to add that generality. Weights would need to be added to the definition of a neighbor bond in the NeighborQuery code. For both `query` and `queryBall`, the weights should always be `1`.

Reply:
I think this would be better done with making NeighborList to contain distance also.

Reply:
Could you elaborate more on adding weights to NeighborQuery?

Reply:
@jinsoo960 The neighbors found by NeighborQuery classes currently include two indices and a distance. I would like to add a `float weight` field to these. For ball queries and k-nearest neighbor queries, the weights would always default to `1.0`.